pyvisa/resources/messagebased.py

```diff
@@ -469,17 +469,44 @@ def read_binary_values(self, datatype='f', is_big_endian=False,
         :rtype: type(container)
 
         """
-        block = self._read_raw()
+        block = b''
         expected_length = 0
 
         if header_fmt == 'ieee':
-            offset, data_length = util.parse_ieee_block_header(block)
+            hash = self.read_bytes(1)                   # Hash '#'
+            digits = self.read_bytes(1)                 # Digits
+            block_size = self.read_bytes(int(digits))   # Block size
+
+            block = block + hash
+            block = block + digits
+            block = block + block_size
+
+            block = block + self.read_bytes(int(block_size))
```

Comment:
We should not read all the data at this stage. It is too early in particular because we did not yet address the question of whether or not the termination character should be read.

Reply:
This comment I do not understand. The previous code would read too early. This one only reads the amount of data it "knows" that should be there. Isn't the termination after that?

Reply:
My point is that no matter the protocol at one point we need to read a block of data. It make more sense to go through the same code for all protocols to actually read the data once we determined the proper size on a protocol per protocol basis. See the modification I suggest to the reading below.